### ISSUE TITLE
Performance experiment: create the speculative gecko session sooner to improve applink (view) startup

### ISSUE THREAD

**Issue by acreskeyMoz** [mozilla-mobile/fenix#15973]
The idea behind the experiment:
Create the gecko engine session as soon as possible in order to improve applink (view) startup time.

Since the view test (measuring navigation start and loadEvent) depends on the gecko engine and the content process if we can instantiate the gecko engine sooner, we may be able to improve our timings.

**Procedure**

Create a speculative gecko engine session in FenixApplication `OnCreate()`
```
components.core.engine.speculativeCreateSession(false)
```

https://github.com/mozilla-mobile/fenix/blob/0e2657e2620457c502e6b79b40f67ced6f8f0117/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#L120

**Results**

No measured difference:

|Variant | Median, navStart | Median, loadEventEnd |
| --- | --- | --- |
| Baseline | 2420 | 5496 |
| Experiment | 2420 | 5526 |

Baseline:
https://docs.google.com/spreadsheets/d/1lu_c0hZGMzMpqGuEb8wtA8hwEucItiy-Uf8gpWUYVmo/edit#gid=0&range=B1:G22

Experiment:
https://docs.google.com/spreadsheets/d/1lu_c0hZGMzMpqGuEb8wtA8hwEucItiy-Uf8gpWUYVmo/edit#gid=0&range=B24:G45

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-8252)

**Comment by acreskeyMoz** [comment]
I'll leave this open for a bit to see if anyone has any ideas.

**Comment by mcomella** [comment]
I don't know what the bottleneck is for VIEW so in theory, I'm not sure this would help. However, it seems like a fair assumption that starting the session as early as possible might help.

I can think of a few possibilities:
- there's a bug so `create speculative session` doesn't work. Were you able to verify the session was created earlier? e.g. maybe it's only creating something in the kotlin layer and not going all the way down to GV
- `components.core.engine.warmUp()` (a few lines later) already does this so there's no real impact (though it doesn't look like it from my brief code search – the first session appears to be created lazily when `loadUrl` is called)
- something is acting as a "barrier" (e.g. like the concurrency ones): you create the session earlier but maybe it doesn't take affect until GV is able to schedule a runnable on the main thread to init something (i.e. after `FenixApp.onCreate` or `HomeActivity.onCreate`) – do we know when the session is created?

**Comment by acreskeyMoz** [comment]
Thank you for the ideas Michael.

I've verified that at least at the JVM layer the speculative engine session has been created (and is used for the subsequent applink).

From looking at logs it looks to me that the bottleneck is in the time it takes to initialize the new Gecko session.
My suspicion is that the content process creation is waiting on the gecko parent process initialization which is still running at the point that we've asked for the new gecko session.

One possible optimization:  it looks like the glean initialization is delaying the Gecko startup.
(I'll try swapping the two as a quick test).

**Comment by acreskeyMoz** [comment]
If I create the speculative session _before_ initializing glean, there is evidence of a 25-30ms improvement to `process launch to navStart`.

https://docs.google.com/spreadsheets/d/1lu_c0hZGMzMpqGuEb8wtA8hwEucItiy-Uf8gpWUYVmo/edit#gid=1996398385&range=B1:R6

Not much of an improvement though.

**Comment by acreskeyMoz** [comment]
I ran a few variations, including disabling Glean (with the idea being that we'd initialize after the pageload).

| Variation | Median processLaunchToNavStart, ms | 90% confidence margin of error +/- ms |
| --- | --- | -- |
| Baseline | 2470 | 5.21  |
| Spec geckosession + deferred Glean initialization | 2431 |  5.36 |
| Glean initialization disabled | 2406 | 12.56 |
| Spec geckosession + Glean initialization disabled | 2389 | 6.19 |

Moto G5, 30 iterations of mozperftest `view`, local machine.

The gains look to be statistically significant but pretty modest given the additional complexity.

**Comment by mcomella** [comment]
Let's review whether or not we want to do this in triage.

**Comment by mcomella** [comment]
@acreskeyMoz Do we know how much time improvement speculative connection improves in isolation? We're not sure why glean would have an impact on this. (we added an item to talk about this in the future of cold startup if you want to talk IRL)

> Glean initialization disabled

MarcLeclair says this is because of the shared pref read. We might be able to have Glean init itself, hold off sending until we set an explicit preference, and set it async.

**Comment by acreskeyMoz** [comment]
> @acreskeyMoz Do we know how much time improvement speculative connection improves in isolation? We're not sure why glean would have an impact on this. (we added an item to talk about this in the future of cold startup if you want to talk IRL)

I couldn't measure _any_ improvement from simply creating the speculative gecko session early on in Fenix init (but after Glean init).

I don't have profile evidence to show why creating it before glean init _does_ help, but I'm guessing that it's because the glean init takes some amount of time and so if we can get a head start on the async gecko session creation, it helps it be ready sooner for the incoming pageload request.

Overall, my view is that the potential rewards from these tweaks are not that great and that the primary bottleneck is elsewhere.

Unfortunately I have my hands full in other areas and don't expect to be able to explore further for a while!

**Comment by cpeterso** [comment]
Moved to bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1807313

Change performed by the [Move to Bugzilla add-on](https://addons.mozilla.org/en-US/firefox/addon/move-to-bugzilla/).
